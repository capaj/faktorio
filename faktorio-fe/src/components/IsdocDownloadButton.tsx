import { Button, ButtonProps } from '@/components/ui/button'
import { trpcClient } from '@/lib/trpcClient'
import { generateIsdocXml } from '@/lib/isdoc/generateIsdocXml'
import { snakeCase } from 'lodash-es'
import { useState } from 'react'

interface IsdocDownloadButtonProps extends ButtonProps {
  invoiceId: string
  variant?: 'default' | 'destructive' | 'outline' | 'secondary' | 'ghost' | 'link'
  size?: 'default' | 'sm' | 'lg' | 'icon'
  children?: React.ReactNode
}

export function IsdocDownloadButton({
  invoiceId,
  variant = 'outline',
  size = 'default',
  children,
  ...props
}: IsdocDownloadButtonProps) {
  const [isDownloading, setIsDownloading] = useState(false)
  
  // Pre-fetch invoice data for better UX
  const { data: invoice } = trpcClient.invoices.getById.useQuery(
    { id: invoiceId },
    { enabled: !!invoiceId }
  )
  
  // Pre-fetch invoicing details
  const { data: invoicingDetails } = trpcClient.invoicingDetails.useQuery()
  
  const handleDownload = () => {
    if (isDownloading || !invoice) return
    
    setIsDownloading(true)
    try {
      // Use a safe check for vat_payer property
      const isVatPayer = Boolean(invoicingDetails && 'vat_payer' in invoicingDetails ? invoicingDetails.vat_payer : false)
      const xml = generateIsdocXml(invoice, isVatPayer)
      
      // Create filename for download
      const baseFileName = `${snakeCase(invoice.your_name ?? '')}-${invoice.number}`
      const fileName = `${baseFileName}.isdoc`
      
      // Create a blob and trigger download
      const blob = new Blob([xml], { type: 'application/xml' })
      const url = URL.createObjectURL(blob)
      const a = document.createElement('a')
      a.href = url
      a.download = fileName
      document.body.appendChild(a)
      a.click()
      URL.revokeObjectURL(url)
      document.body.removeChild(a)
    } catch (error) {
      console.error('Error downloading ISDOC:', error)
    } finally {
      setIsDownloading(false)
    }
  }
  
  return (
    <Button 
      variant={variant} 
      size={size} 
      onClick={handleDownload}
      disabled={isDownloading || !invoice}
      {...props}
    >
      {isDownloading ? 'Stahování...' : children || 'Stáhnout ISDOC'}
    </Button>
  )
}
