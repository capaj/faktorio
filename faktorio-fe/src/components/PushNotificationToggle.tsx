import { useState, useEffect } from 'react'
import { Button } from '@/components/ui/button'
import { Bell, BellOff } from 'lucide-react'
import { pushNotificationService } from '@/lib/pushNotificationService'
import { toast } from 'sonner'

export function PushNotificationToggle() {
  const [isSubscribed, setIsSubscribed] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const [isSupported, setIsSupported] = useState(false)

  useEffect(() => {
    async function initializeAndCheck() {
      try {
        // First initialize the service worker
        const supported = await pushNotificationService.initialize()
        setIsSupported(supported)

        // Only check subscription if supported
        if (supported) {
          const subscribed = await pushNotificationService.isSubscribed()
          setIsSubscribed(subscribed)
          console.log('Initial subscription state:', subscribed)
        }
      } catch (error) {
        console.error('Error initializing push notifications:', error)
        setIsSupported(false)
        setIsSubscribed(false)
      }
    }

    initializeAndCheck()
  }, [])

  const handleToggle = async () => {
    if (!isSupported) {
      toast.error('Push notifications are not supported in this browser')
      return
    }

    setIsLoading(true)
    try {
      if (isSubscribed) {
        await pushNotificationService.unsubscribe()
        setIsSubscribed(false)
        toast.success('Notifikace o splatných fakturách byly vypnuty')
      } else {
        await pushNotificationService.subscribe()
        setIsSubscribed(true)
        toast.success('Notifikace o splatných fakturách byly zapnuty')
      }
    } catch (error) {
      console.error('Error toggling notification subscription:', error)

      if (error instanceof Error) {
        if (error.message.includes('permission')) {
          toast.error(
            'Povolení pro notifikace bylo zamítnuto. Povolte notifikace v nastavení prohlížeče.'
          )
        } else {
          toast.error(
            'Nastala chyba při nastavování notifikací: ' + error.message
          )
        }
      } else {
        toast.error('Nastala chyba při nastavování notifikací')
      }
    } finally {
      setIsLoading(false)
    }
  }

  if (!isSupported) {
    return null // Don't render anything if not supported
  }

  return (
    <Button
      variant={isSubscribed ? 'default' : 'outline'}
      size="sm"
      onClick={handleToggle}
      disabled={isLoading}
      className="flex items-center gap-2"
    >
      {isSubscribed ? (
        <>
          <Bell className="h-4 w-4" />
          Notifikace zapnuty
        </>
      ) : (
        <>
          <BellOff className="h-4 w-4" />
          Zapnout notifikace
        </>
      )}
    </Button>
  )
}
