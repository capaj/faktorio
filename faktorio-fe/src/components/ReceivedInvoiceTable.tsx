import React from 'react'
import {
  TableHeader,
  TableRow,
  TableHead,
  TableBody,
  TableCell,
  TableFooter,
  Table
} from '@/components/ui/table'
import { Button } from '@/components/ui/button'
import { Trash2Icon, Loader2 } from 'lucide-react' // Keep used icons

// Define the shape of a received invoice
import { receivedInvoiceTb } from 'faktorio-db/schema'
import { InferSelectModel } from 'drizzle-orm'

export type ReceivedInvoice = Pick<
  InferSelectModel<typeof receivedInvoiceTb>,
  | 'id'
  | 'supplier_name'
  | 'supplier_vat_no'
  | 'invoice_number'
  | 'issue_date'
  | 'taxable_supply_date'
  | 'due_date'
  | 'total_without_vat'
  | 'total_with_vat'
  | 'currency'
  | 'status'
>

interface ReceivedInvoiceTableProps {
  invoices: ReceivedInvoice[]
  columns?: {
    textAlign?: string
    name: string
    label: string
  }[]
  onDelete?: (id: string) => Promise<void> // Optional delete handler
  isDeleting?: boolean // Optional flag for delete loading state
  deletingId?: string | null // Optional ID of invoice being deleted
  showTotals?: boolean
}

const defaultColumns = [
  {
    name: 'supplier_name',
    label: 'Dodavatel'
  },
  {
    name: 'invoice_number',
    label: 'Číslo faktury'
  },
  {
    name: 'issue_date',
    label: 'Datum vystavení'
  },
  {
    name: 'due_date',
    label: 'Datum splatnosti'
  },
  {
    name: 'vat_amount',
    label: 'Částka DPH',
    textAlign: 'right'
  },
  {
    name: 'total_without_vat',
    label: 'Celkem bez DPH',
    textAlign: 'right'
  },
  {
    name: 'total_with_vat',
    label: 'Celkem s DPH',
    textAlign: 'right'
  }
  // {
  //   name: 'status',
  //   label: 'Stav'
  // }
]

export function ReceivedInvoiceTable({
  invoices,
  columns = defaultColumns,
  onDelete,
  isDeleting,
  deletingId
}: ReceivedInvoiceTableProps) {
  // Calculate totals
  const totalWithoutVatSum = invoices.reduce(
    (sum, inv) => sum + (inv.total_without_vat ?? 0),
    0
  )
  const totalWithVatSum = invoices.reduce(
    (sum, inv) => sum + inv.total_with_vat,
    0
  )
  // Determine currency (assuming all invoices in the list have the same currency)
  const currency = invoices.length > 0 ? invoices[0].currency : ''

  const formatNumber = (num: number | null | undefined) => {
    if (num === null || num === undefined) return '-'
    return num.toLocaleString() // Simple localization for now
  }

  const totalVatSum = invoices.reduce(
    (sum, inv) => sum + (inv.total_with_vat - (inv.total_without_vat ?? 0)),
    0
  )

  return (
    <Table>
      <TableHeader>
        <TableRow>
          {columns?.map((column) => (
            <TableHead
              className={column.textAlign === 'right' ? 'text-right' : ''}
              key={column.name}
            >
              {column.label}
            </TableHead>
          ))}
          {onDelete && <TableHead className="text-right">Akce</TableHead>}{' '}
          {/* Show actions only if handler provided */}
        </TableRow>
      </TableHeader>
      <TableBody>
        {invoices.length === 0 && (
          <TableRow>
            <TableCell colSpan={onDelete ? 8 : 7} className="text-center">
              Žádné přijaté faktury k zobrazení.
            </TableCell>
          </TableRow>
        )}
        {invoices.map((invoice) => (
          <TableRow key={invoice.id}>
            <TableCell className="font-medium">
              {invoice.supplier_name}
            </TableCell>
            <TableCell>{invoice.invoice_number}</TableCell>
            <TableCell>{invoice.issue_date}</TableCell>
            <TableCell>{invoice.due_date}</TableCell>
            <TableCell className="text-right">
              {formatNumber(
                invoice.total_with_vat - (invoice.total_without_vat ?? 0)
              )}
              {invoice.currency}
            </TableCell>
            <TableCell className="text-right">
              {formatNumber(invoice.total_without_vat)} {invoice.currency}
            </TableCell>
            <TableCell className="text-right">
              {formatNumber(invoice.total_with_vat)} {invoice.currency}
            </TableCell>

            {onDelete && (
              <TableCell className="text-right">
                <Button
                  variant="ghost"
                  size="icon"
                  onClick={() => onDelete(invoice.id)}
                  disabled={isDeleting && deletingId === invoice.id}
                  className="text-red-600 hover:text-red-800 disabled:text-gray-400"
                  aria-label="Smazat fakturu"
                >
                  {isDeleting && deletingId === invoice.id ? (
                    <Loader2 className="h-4 w-4 animate-spin" />
                  ) : (
                    <Trash2Icon className="h-4 w-4" />
                  )}
                </Button>
              </TableCell>
            )}
          </TableRow>
        ))}
      </TableBody>
      {invoices.length > 0 && (
        <TableFooter>
          <TableRow>
            <TableCell colSpan={4} className="font-medium">
              Celkem
            </TableCell>
            <TableCell className="text-right font-medium">
              {formatNumber(totalVatSum)} {currency}
            </TableCell>
            <TableCell className="text-right font-medium">
              {formatNumber(totalWithoutVatSum)} {currency}
            </TableCell>
            <TableCell className="text-right font-medium">
              {formatNumber(totalWithVatSum)} {currency}
            </TableCell>
            {/* Adjust colspan based on whether actions column is present */}
            <TableCell colSpan={onDelete ? 2 : 1}></TableCell>
          </TableRow>
        </TableFooter>
      )}
    </Table>
  )
}
