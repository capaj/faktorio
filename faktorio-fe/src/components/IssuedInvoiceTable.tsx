import React from 'react'
import {
  TableCaption,
  TableHeader,
  TableRow,
  TableHead,
  TableBody,
  TableCell,
  Table
} from '@/components/ui/table'
import {
  DropdownMenu,
  DropdownMenuTrigger,
  DropdownMenuContent,
  DropdownMenuItem
} from '@/components/ui/dropdown-menu'
import { Button } from '@/components/ui/button'
import { RemoveDialogUncontrolled } from '@/components/RemoveDialog'
import {
  LucideEllipsisVertical,
  Pencil,
  Trash2,
  CheckCircle,
  XCircle
} from 'lucide-react'
import { Link } from 'wouter'
import { formatNumberWithSpaces } from '@/pages/formatNumberWithSpaces' // Adjust path if needed
import { InvoicesDownloadButton } from '@/pages/InvoiceList/InvoicesDownloadButton' // Adjust path if needed

// Define the shape of an invoice - adjust based on your actual data structure
// This might need refinement based on the exact structure from your tRPC query
import { invoicesTb } from 'faktorio-db/schema'
import { InferSelectModel } from 'drizzle-orm'

export type Invoice = Pick<
  InferSelectModel<typeof invoicesTb>,
  | 'id'
  | 'number'
  | 'client_name'
  | 'taxable_fulfillment_due'
  | 'issued_on'
  | 'sent_at'
  | 'total'
  | 'native_total'
  | 'native_subtotal'
  | 'subtotal'
  | 'currency'
  | 'paid_on'
  | 'client_vat_no'
  | 'exchange_rate'
>

// Helper type for currency totals
type CurrencyTotals = {
  [currency: string]: {
    total: number
    subtotal: number
    count: number
  }
}

interface InvoiceTableProps {
  invoices: Invoice[]
  isLoading: boolean
  onDelete?: (id: string) => Promise<void>
  onMarkAsPaid?: (id: string, number: string) => void
  onMarkAsUnpaid?: (id: string) => Promise<void>
  showTotals?: boolean
  year?: number | null
  search?: string
  // Add any other props needed, e.g., specific mutation status if needed inside the table
}

export function IssuedInvoiceTable({
  invoices,
  isLoading,
  onDelete,
  onMarkAsPaid,
  onMarkAsUnpaid,
  showTotals = true, // Default to showing totals
  year,
  search
}: InvoiceTableProps) {
  // Determine if we should show full year in dates
  const showFullYear = year === null || year === undefined

  // Helper function to format dates in Czech format
  function formatCzechDate(dateString: string | null): string {
    if (!dateString) return ''

    const date = new Date(dateString)
    const day = date.getDate()
    const month = date.getMonth() + 1
    const year = date.getFullYear()

    if (showFullYear) {
      return `${day}.${month}.${year}`
    } else {
      return `${day}.${month}.`
    }
  }

  const currencyTotals = invoices.reduce<CurrencyTotals>((acc, invoice) => {
    const currency = invoice.currency || 'N/A' // Handle potential null/undefined currency
    if (!acc[currency]) {
      acc[currency] = { total: 0, subtotal: 0, count: 0 }
    }
    acc[currency].total += invoice.total
    acc[currency].subtotal += invoice.subtotal ?? 0
    acc[currency].count += 1
    return acc
  }, {})

  // Calculate total sum converted to CZK
  const totalSumCZK = invoices.reduce((acc, invoice) => {
    let amountInCZK = invoice.total
    if (invoice.currency !== 'CZK' && invoice.exchange_rate) {
      amountInCZK = invoice.total * invoice.exchange_rate
    } else if (invoice.currency !== 'CZK') {
      // Handle cases where exchange rate might be missing for non-CZK invoices
      // For now, we'll skip them in the total sum, but you might want a different handling
      console.warn(
        `Missing exchange rate for non-CZK invoice ${invoice.number}`
      )
      return acc
    }
    return acc + amountInCZK
  }, 0)
  const totalVAT = invoices.reduce((acc, invoice) => {
    // Assuming native_total includes VAT, we can calculate VAT as:
    const vat = (invoice.native_total ?? 0) - invoice.native_subtotal
    return acc + (vat > 0 ? vat : 0) // Only add positive VAT amounts
  }, 0)

  const sortedCurrencies = Object.keys(currencyTotals).sort()

  return (
    <Table>
      <TableHeader className="bg-gray-50">
        <TableRow>
          <TableHead className="w-[100px]">Nr.</TableHead>
          <TableHead>Klient</TableHead>
          <TableHead>Zdanitelné plnění datum</TableHead>
          <TableHead>Vystaveno dne</TableHead>
          <TableHead>Odesláno</TableHead>
          <TableHead>Datum platby</TableHead>
          <TableHead>bez DPH</TableHead>
          <TableHead>s DPH</TableHead>
          <TableHead className="text-right">Akce</TableHead>
        </TableRow>
      </TableHeader>
      <TableBody>
        {invoices.map((invoice) => (
          <TableRow key={invoice.id}>
            <TableCell className="font-medium">
              <Link href={`/invoices/${invoice.id}/?language=cs`}>
                {invoice.number}
              </Link>
            </TableCell>
            <TableCell>{invoice.client_name}</TableCell>
            <TableCell>
              {formatCzechDate(invoice.taxable_fulfillment_due)}
            </TableCell>
            <TableCell>{formatCzechDate(invoice.issued_on)}</TableCell>
            <TableCell>{formatCzechDate(invoice.sent_at)}</TableCell>
            <TableCell>
              <span
                className={`px-2 py-1 rounded-full text-xs font-medium ${
                  invoice.paid_on
                    ? 'bg-green-100 text-green-800'
                    : 'bg-yellow-100 text-yellow-800'
                }`}
              >
                {invoice.paid_on
                  ? formatCzechDate(invoice.paid_on)
                  : 'Nezaplaceno'}
              </span>
            </TableCell>
            <TableCell>
              {invoice.subtotal} {invoice.currency}
              {invoice.currency !== 'CZK' &&
                invoice.exchange_rate &&
                invoice.subtotal && (
                  <div className="text-xs text-gray-500">
                    ({invoice.exchange_rate * invoice.subtotal} CZK)
                  </div>
                )}
            </TableCell>
            <TableCell>
              {invoice.total} {invoice.currency}
            </TableCell>

            <TableCell className="text-right">
              <DropdownMenu>
                <DropdownMenuTrigger asChild>
                  <Button variant="ghost" className="bg-gray-200">
                    <LucideEllipsisVertical className="h-5 w-5 text-gray-400" />
                  </Button>
                </DropdownMenuTrigger>
                <DropdownMenuContent align="end" className="cursor-pointer">
                  <DropdownMenuItem>
                    <Link
                      href={`/invoices/${invoice.id}/edit`}
                      className="flex w-full"
                    >
                      <Pencil size={16} strokeWidth="1.5" />
                      <span className="ml-2">Editovat</span>
                    </Link>
                  </DropdownMenuItem>
                  <DropdownMenuItem>
                    <Link
                      href={`/invoices/${invoice.id}`}
                      className="flex w-full"
                    >
                      <Pencil size={16} strokeWidth="1.5" />
                      <span className="ml-2">Zobrazit PDF</span>
                    </Link>
                  </DropdownMenuItem>

                  {!invoice.paid_on && onMarkAsPaid && (
                    <DropdownMenuItem
                      className="cursor-pointer"
                      onSelect={(e) => e.preventDefault()}
                      onClick={() => onMarkAsPaid(invoice.id, invoice.number)}
                    >
                      <CheckCircle
                        size={16}
                        strokeWidth="1.5"
                        className="text-green-600"
                      />
                      <span className="ml-2">Označit jako zaplacené</span>
                    </DropdownMenuItem>
                  )}

                  {invoice.paid_on && onMarkAsUnpaid && (
                    <DropdownMenuItem
                      className="text-red-600 focus:text-red-700 focus:bg-red-50"
                      onSelect={(e) => e.preventDefault()}
                      onClick={async () => {
                        await onMarkAsUnpaid(invoice.id)
                      }}
                      // Consider passing mutation status if needed for disabling
                      // disabled={markAsUnpaidMutation.isPending}
                    >
                      <XCircle size={16} strokeWidth="1.5" />
                      <span className="ml-2">Označit jako nezaplacené</span>
                    </DropdownMenuItem>
                  )}

                  {onDelete && (
                    <RemoveDialogUncontrolled
                      title={
                        <span>
                          Opravdu chcete smazat fakturu{' '}
                          <strong>{invoice.number}</strong>?
                        </span>
                      }
                      onRemove={async () => {
                        await onDelete(invoice.id)
                      }}
                    >
                      <DropdownMenuItem className="cursor-pointer">
                        <Trash2 size={16} strokeWidth="1.5" />
                        <span className="ml-2">Smazat</span>
                      </DropdownMenuItem>
                    </RemoveDialogUncontrolled>
                  )}
                </DropdownMenuContent>
              </DropdownMenu>
            </TableCell>
          </TableRow>
        ))}
        {isLoading && (
          <TableRow>
            <TableCell colSpan={9} className="text-center">
              Načítám faktury...
            </TableCell>
          </TableRow>
        )}
        {!isLoading && invoices.length === 0 && (
          <TableRow>
            <TableCell colSpan={9} className="text-center">
              Žádné faktury k zobrazení.
            </TableCell>
          </TableRow>
        )}

        {showTotals && !isLoading && invoices.length > 0 && (
          <>
            {/* Dynamically render rows for each currency */}
            {sortedCurrencies.map((currency) => (
              <TableRow key={currency} className="bg-gray-100 font-medium">
                <TableCell colSpan={5}>
                  Celkem {currencyTotals[currency].count}{' '}
                  {currencyTotals[currency].count === 1
                    ? 'faktura'
                    : currencyTotals[currency].count > 1 &&
                        currencyTotals[currency].count < 5
                      ? 'faktury'
                      : 'faktur'}{' '}
                  v {currency}
                </TableCell>
                <TableCell></TableCell> {/* Empty cell for alignment */}
                <TableCell className="whitespace-nowrap">
                  {formatNumberWithSpaces(currencyTotals[currency].subtotal)}{' '}
                  {currency}
                </TableCell>
                <TableCell className="whitespace-nowrap">
                  {formatNumberWithSpaces(currencyTotals[currency].total)}{' '}
                  {currency}
                </TableCell>
                <TableCell className="text-right"></TableCell>{' '}
                {/* Empty action cell */}
              </TableRow>
            ))}

            {/* Overall Total Count, Sum, and Download Row - Only show if multiple currencies */}
            {Object.keys(currencyTotals).length > 1 && (
              <TableRow className="bg-gray-200 font-semibold">
                <TableCell colSpan={3}>
                  {' '}
                  {/* Adjusted colspan to push sum next to button */}
                  Ve všech měnách {invoices.length}{' '}
                  {invoices.length === 1
                    ? 'faktura'
                    : invoices.length > 1 && invoices.length < 5
                      ? 'faktury'
                      : 'faktur'}
                </TableCell>
                <TableCell colSpan={3} className="text-right font-light">
                  Celkem částka DPH:{' '}
                  <span className="font-bold text-yellow-800">
                    {formatNumberWithSpaces(totalVAT)} CZK
                  </span>
                </TableCell>

                <TableCell colSpan={1} className="text-left whitespace-nowrap">
                  {formatNumberWithSpaces(
                    invoices.reduce((acc, inv) => acc + inv.native_subtotal, 0)
                  )}{' '}
                  CZK
                </TableCell>
                {/* Display the total sum in CZK */}
                <TableCell colSpan={1} className="text-left whitespace-nowrap">
                  {formatNumberWithSpaces(totalSumCZK)} CZK
                </TableCell>

                <TableCell className="text-right">
                  {year !== undefined && search !== undefined && (
                    <InvoicesDownloadButton year={year} search={search} />
                  )}
                </TableCell>
              </TableRow>
            )}
          </>
        )}
      </TableBody>
    </Table>
  )
}
