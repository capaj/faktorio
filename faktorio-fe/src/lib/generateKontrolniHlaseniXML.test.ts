import { describe, it, expect, vi, afterEach } from 'vitest'
import {
  generateKontrolniHlaseniXML,
  type SubmitterData
} from './generateKontrolniHlaseniXML'

import { ReceivedInvoice } from '@/components/ReceivedInvoiceTable'
import { Invoice } from '@/components/IssuedInvoiceTable'

describe('generateKontrolniHlaseniXML', () => {
  afterEach(() => {
    vi.useRealTimers()
  })

  it('should generate correct XML for a simple happy path', () => {
    vi.useFakeTimers()
    vi.setSystemTime(new Date('2025-04-21'))

    const mockIssuedInvoices: Invoice[] = [
      {
        id: 'iss1',
        number: '2024-001',
        client_name: 'Client A',
        client_vat_no: 'CZ11111111', // Added mock VAT ID
        taxable_fulfillment_due: '2024-07-15',
        issued_on: '2024-07-10',
        sent_at: '2024-07-11',
        due_on: '2024-07-20',
        total: 12100,
        subtotal: 10000,
        native_subtotal: 10000, // Use native_subtotal for consistency
        native_total: 12100, // 21% VAT
        currency: 'CZK',
        exchange_rate: 1,
        paid_on: '2024-07-20'
      }
    ]

    const mockReceivedInvoices: ReceivedInvoice[] = [
      {
        id: 'rec1',
        supplier_name: 'Supplier B',
        supplier_vat_no: 'CZ22222222', // Added mock VAT ID
        invoice_number: 'INV-B-100',
        issue_date: '2024-08-01',
        taxable_supply_date: '2024-08-01',
        due_date: '2024-08-15',
        total_without_vat: 5000,
        total_with_vat: 6050, // Below threshold
        currency: 'CZK',
        status: 'paid'
      },
      {
        id: 'rec2',
        supplier_name: 'Supplier C',
        supplier_vat_no: 'CZ33333333', // Added mock VAT ID
        invoice_number: 'INV-C-200',
        issue_date: '2024-08-05',
        taxable_supply_date: '2024-08-05',
        due_date: '2024-08-20',
        total_without_vat: 15000,
        total_with_vat: 18150, // Above threshold
        currency: 'CZK',
        status: 'received'
      },
      {
        id: 'rec3',
        supplier_name: 'Supplier D (EUR)',
        supplier_vat_no: 'DE44444444', // Added mock VAT ID
        invoice_number: 'INV-D-300',
        issue_date: '2024-08-10',
        taxable_supply_date: '2024-08-10',
        due_date: '2024-08-25',
        total_without_vat: 100,
        total_with_vat: 121,
        currency: 'EUR',
        status: 'received'
      }
    ]

    const mockSubmitterData: SubmitterData = {
      dic: 'CZ12345678',
      naz_obce: 'Brno',
      typ_ds: 'F',
      jmeno: 'Test',
      prijmeni: 'Submitter',
      ulice: 'Test Street 123',
      psc: '12345',
      stat: 'ČESKÁ REPUBLIKA',
      email: 'test@submitter.com'
    }

    const year = 2024
    const quarter = 3

    // --- Generate XML ---
    const xmlString = generateKontrolniHlaseniXML({
      issuedInvoices: mockIssuedInvoices,
      receivedInvoices: mockReceivedInvoices,
      submitterData: mockSubmitterData,
      year,
      quarter
    })

    expect(xmlString).toMatchSnapshot()
  })

  it('should handle reverse charge invoice inside czechia-czech contractor invoicing a czech company', () => {
    // this is a special case where the invoice is issued by a czech company to a czech company, but it is a reverse charge invoice. It only applies to a very limited set of goods and services, for example https://financnisprava.gov.cz/cs/financni-sprava/media-a-verejnost/tiskove-zpravy-gfr/tiskove-zpravy-2017/od-cervence-dochazi-k-rozsireni-rezimu-reverse-charge-na-dalsi-plneni

    vi.useFakeTimers()
    vi.setSystemTime(new Date('2025-04-21'))

    const mockIssuedInvoices: Invoice[] = [
      {
        id: 'iss1',
        number: '2024-001',
        client_name: 'Client A',
        client_vat_no: 'CZ11111111', //
        taxable_fulfillment_due: '2024-07-15',
        issued_on: '2024-07-10',
        sent_at: '2024-07-11',
        due_on: '2024-07-20',
        total: 12100,
        subtotal: 12100,
        native_subtotal: 12100, // Use native_subtotal for consistency
        native_total: 12100, // 0% VAT for reverse charge
        currency: 'CZK',
        exchange_rate: 1,
        paid_on: '2024-07-20'
      }
    ]

    const mockReceivedInvoices: ReceivedInvoice[] = [
      {
        id: 'rec1',
        supplier_name: 'Supplier B',
        supplier_vat_no: 'CZ22222222', //
        invoice_number: 'INV-B-100',
        issue_date: '2024-08-01',
        taxable_supply_date: '2024-08-01',
        due_date: '2024-08-15',
        total_without_vat: 5000,
        total_with_vat: 6050, // Below threshold
        currency: 'CZK',
        status: 'paid'
      }
    ]

    const mockSubmitterData: SubmitterData = {
      dic: 'CZ12345678',
      naz_obce: 'Brno',
      typ_ds: 'F',
      jmeno: 'Test',
      prijmeni: 'Submitter',
      ulice: 'Test Street 123',
      psc: '12345',
      stat: 'ČESKÁ REPUBLIKA',
      email: 'test@submitter.com'
    }
    const year = 2024
    const month = 7 // July
    // --- Generate XML ---
    const xmlString = generateKontrolniHlaseniXML({
      issuedInvoices: mockIssuedInvoices,
      receivedInvoices: mockReceivedInvoices,
      submitterData: mockSubmitterData,
      year,
      month
    })

    expect(xmlString).toMatchSnapshot()
  })

  it('should include credit notes (dobropisy) with negative values in VetaB2 when above the threshold', () => {
    vi.useFakeTimers()
    vi.setSystemTime(new Date('2025-02-15'))

    const mockIssuedInvoices: Invoice[] = []

    const mockReceivedInvoices: ReceivedInvoice[] = [
      {
        id: 'rec-positive',
        supplier_name: 'Dodavatel s.r.o.',
        supplier_vat_no: 'CZ11111111',
        invoice_number: 'F2025-0001',
        issue_date: '2025-01-15',
        taxable_supply_date: '2025-01-15',
        due_date: '2025-01-29',
        total_without_vat: 52000,
        total_with_vat: 62920,
        currency: 'CZK',
        status: 'paid'
      },
      {
        id: 'rec-credit-note',
        supplier_name: 'Dodavatel s.r.o.',
        supplier_vat_no: 'CZ11111111',
        invoice_number: 'DN2025-0001',
        issue_date: '2025-02-10',
        taxable_supply_date: '2025-02-10',
        due_date: '2025-02-24',
        total_without_vat: -26000,
        total_with_vat: -31460,
        currency: 'CZK',
        status: 'received'
      }
    ]

    const mockSubmitterData: SubmitterData = {
      dic: 'CZ76543210',
      naz_obce: 'Praha',
      typ_ds: 'F',
      jmeno: 'Jan',
      prijmeni: 'Novák',
      ulice: 'Hlavní 1',
      psc: '11000',
      stat: 'ČESKÁ REPUBLIKA',
      email: 'jan.novak@example.com'
    }

    const xmlString = generateKontrolniHlaseniXML({
      issuedInvoices: mockIssuedInvoices,
      receivedInvoices: mockReceivedInvoices,
      submitterData: mockSubmitterData,
      year: 2025,
      month: 2
    })

    expect(xmlString).toContain(
      'c_evid_dd="F2025-0001"\n      dppd="15.01.2025"\n      zakl_dane1="52000"\n      dan1="10920"'
    )

    expect(xmlString).toContain(
      'c_evid_dd="DN2025-0001"\n      dppd="10.02.2025"\n      zakl_dane1="-26000"\n      dan1="-5460"'
    )

    expect(xmlString).toContain('pln23="26000"')
  })

  it('reports received invoices in VetaB2 when only the VAT-inclusive total exceeds the CZK 10 000 threshold', () => {
    const mockReceivedInvoices: ReceivedInvoice[] = [
      {
        id: 'rec-threshold',
        supplier_name: 'Dodavatel s.r.o.',
        supplier_vat_no: 'CZ11112222',
        invoice_number: 'F2025-0002',
        issue_date: '2025-02-01',
        taxable_supply_date: '2025-02-01',
        due_date: '2025-02-15',
        total_without_vat: 9000,
        total_with_vat: 10890,
        currency: 'CZK',
        status: 'paid'
      }
    ]

    const xmlString = generateKontrolniHlaseniXML({
      issuedInvoices: [],
      receivedInvoices: mockReceivedInvoices,
      submitterData: {
        dic: 'CZ76543210',
        naz_obce: 'Praha',
        typ_ds: 'F',
        jmeno: 'Jan',
        prijmeni: 'Novák',
        ulice: 'Hlavní 1',
        psc: '11000',
        stat: 'ČESKÁ REPUBLIKA',
        email: 'jan.novak@example.com'
      },
      year: 2025,
      month: 2
    })

    expect(xmlString).toContain(
      'dic_dod="11112222"\n      c_evid_dd="F2025-0002"\n      dppd="01.02.2025"\n      zakl_dane1="9000"\n      dan1="1890"'
    )
  })
})
