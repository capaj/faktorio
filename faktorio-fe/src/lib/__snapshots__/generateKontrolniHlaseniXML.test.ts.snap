// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateKontrolniHlaseniXML > should generate correct XML for a simple happy path 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<Pisemnost nazevSW="EPO MF ČR" verzeSW="41.16.3">
<DPHKH1 verzePis="03.01">
  <VetaD k_uladis="DPH" dokument="KH1"
    rok="2024" ctvrt="3"
    d_poddp="21.04.2025"
    khdph_forma="B"
  />
  <VetaP
    dic="12345678" typ_ds="F" jmeno="Test" prijmeni="Submitter" ulice="Test Street 123" psc="12345" stat="ČESKÁ REPUBLIKA" email="test@submitter.com" sest_jmeno="Test" sest_prijmeni="Submitter"
  />
  
  
    <VetaA4
      c_radku="1"
      dic_odb="11111111"
      c_evid_dd="2024-001"
      dppd="15.07.2024"
      zakl_dane1="10000"
      dan1="2100"
      kod_rezim_pl="0"
      zdph_44="N"
    />
  
    <VetaB2
      c_radku="1"
      dic_dod="33333333"
      c_evid_dd="INV-C-200"
      dppd="05.08.2024"
      zakl_dane1="15000"
      dan1="3150"
      zdph_44="N" 
      pomer="N"
    />
  
    <VetaB3
      zakl_dane1="5100"
      dan1="1071"
    />
  <VetaC
    obrat23="10000"
    pln23="20100"
  />
</DPHKH1>
</Pisemnost>"
`;

exports[`generateKontrolniHlaseniXML > should handle reverse charge invoice inside czechia-czech contractor invoicing a czech company 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<Pisemnost nazevSW="EPO MF ČR" verzeSW="41.16.3">
<DPHKH1 verzePis="03.01">
  <VetaD k_uladis="DPH" dokument="KH1"
    rok="2024" mesic="7"
    d_poddp="21.04.2025"
    khdph_forma="B"
  />
  <VetaP
    dic="12345678" typ_ds="F" jmeno="Test" prijmeni="Submitter" ulice="Test Street 123" psc="12345" stat="ČESKÁ REPUBLIKA" email="test@submitter.com" sest_jmeno="Test" sest_prijmeni="Submitter"
  />
  
    <VetaA1
      c_radku="1"
      dic_odb="11111111"
      c_evid_dd="2024-001"
      duzp="15.07.2024"
      zakl_dane1="12100"
    />
  
  
  
    <VetaB3
      zakl_dane1="5000"
      dan1="1050"
    />
  <VetaC
    obrat23="12100"
    pln23="5000"
  />
</DPHKH1>
</Pisemnost>"
`;
